ElevatorSim/src/elevatorsim/scheduler/SchedulerServer.java

```diff
@@ -45,37 +54,102 @@ public Role getRole() {
+			ArrayList<ElevatorRequest> storedRequests = Scheduler.getInstance().getStoredRequests();
+			storedRequests.add(elevatorRequest);
+			InetAddress elevatorAddress = Scheduler.getInstance().findAvailableElevator();
+			ElevatorStatus elevatorStatus = Scheduler.getInstance().getElevators().get(elevatorAddress);
+			if( storedRequests.size() == 1 && elevatorStatus.getDirection() == null ) {
+				elevatorStatus.addFloor(elevatorRequest.getStartFloor());
+				
+				elevatorStatus.setDirection( elevatorRequest.getStartFloor() - elevatorStatus.getFloor() > 0 ? Direction.UP : Direction.DOWN );
+				sender.send(MessagePackets.generateElevatorStateChange(new ElevatorStateChange(ElevatorState.DOOR_CLOSED)), elevatorAddress, elevators.get(elevatorAddress));
+			}
+			return true;
+		}
+		System.out.println();
+		
+		
+		int elevatorStartFloor = Scheduler.getInstance().getElevators().get(availableElevator).getFloor();
+		int boardFloor = elevatorRequest.getStartFloor();
+		int floorChange = boardFloor - elevatorStartFloor;
+		Direction direction;
+		if( floorChange < 0 ) {
+			direction = Direction.DOWN;
+		} else if (floorChange > 0) {
+			direction = Direction.UP;
+		}else {
+			direction = null;
+		}
+		if (direction == null) {
+			//send request to elevator so it can send back it's version
+			Integer elevatorPort = elevators.get(availableElevator);
+			sender.send(MessagePackets.generateElevatorRequest(elevatorRequest), availableElevator, elevatorPort);
+		}
+	
+		return true;
+	}
+	
 	@Override
 	public DatagramPacket handleElevatorRequest(DatagramPacket request) {
 		Scheduler.getInstance().startProcessing();
-
-		InetAddress availableElevator = Scheduler.getInstance().findAvailableElevator();
-
-		if (availableElevator == null) {
-			Scheduler.getInstance().stopProcessing();
-			return MessagePackets.Responses.RESPONSE_FAILURE();
+		ElevatorRequest elevatorRequest = MessagePackets.deserializeElevatorRequest(request.getData());
+		System.out.print("SchedulerServer - Info: Received elevator request " + elevatorRequest.toString() + "\n");
+		boolean success = true;
+		if (elevatorRequest.getStartFloor() == null) {
+			
+				//ugly way of adding floor request to only elevator - will improve in future iteration
+			ElevatorStatus elevatorStatus = Scheduler.getInstance().getElevators().get( Scheduler.getInstance().findAvailableElevator());
+				
+			Direction newDirection = elevatorStatus.addFloor( elevatorRequest.getDestFloor() );
+			if (newDirection != null) {
+				new java.util.Timer().schedule( 
+			        new java.util.TimerTask() {
+			            @Override
+			            public void run() {
+			            	Scheduler.getInstance().startProcessing();
+			                sender.send(DatagramPacketUtils.getCopyOf( MessagePackets.generateElevatorStateChange(new ElevatorStateChange( ElevatorState.DOOR_CLOSED ))) , request.getAddress(), elevators.get(request.getAddress()));
```

Comment:
you don't need to get the copy of a generated packet. you only need the copy when forwarding an existing packet

Reply:
stopped copying